README.md
# sped-nfse-equiplano

# BETHA TESTES

## Este pacote atende o provedor Equiplano

[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]

[![Latest Stable Version][ico-stable]][link-packagist]
[![Latest Version on Packagist][ico-version]][link-packagist]
[![License][ico-license]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]

[![Issues][ico-issues]][link-issues]
[![Forks][ico-forks]][link-forks]
[![Stars][ico-stars]][link-stars]


## Municípios atendidos pelo provedor

|n|Município|UF|Ibge|Entidade|
|:---:|:---|:---:|:---:|:---:|
|1|Alvorada do Sul|PR|4100806|350|
|2|Balsa Nova|PR|4102307|23|
|3|Barra do Jacaré|PR|4102703|29|
|4|Cafelândia|PR|4103453|57|
|5|Cândido de Abreu|PR|4104402|108|
|6|Candói|PR|4104428|7|
|7|Cantagalo|PR|4104451|51|
|8|Capanema|PR|4104501|50|
|9|Conselheiro Mairinck|PR|4106100|22|
|10|Cruzeiro do Iguaçu|PR|4106571|417|
|11|Dois Vizinhos|PR|4107207|68|
|12|Espigão Alto do Iguaçu|PR|4107546|131|
|13|Fernandes Pinheiro|PR|4107736|140|
|14|Francisco Beltrão|PR|4108403|35|
|15|Ibaiti|PR|4109708|3|
|16|Imbituva|PR|4110102|52|
|17|Inácio Martins|PR|4110201|88|
|18|Itaperuçu|PR|4111258|79|
|19|Ivaí|PR|4111407|20|
|20|Jaboti|PR|4111704|59|
|21|Japira|PR|4112306|44|
|22|Jataizinho|PR|4112702|384|
|23|Jundiaí do Sul|PR|4112900|41|
|24|Laranjeiras do Sul|PR|4113304|53|
|25|Manfrinópolis|PR|4114351|150|
|26|Missal|PR|4116059|12|
|27|Nova Laranjeiras|PR|4117057|76|
|28|Ouro Verde do Oeste|PR|4117453|197|
|29|Palmital|PR|4117800|38|
|30|Pinhal de São Bento|PR|4119251|80|
|31|Planalto|PR|4119806|30|
|32|Porto Barreiro|PR|4120150|132|
|33|Pranchita|PR|4120358|47|
|34|Quedas do Iguaçu|PR|4120903|6|
|35|Realeza|PR|4121406|49|
|36|Ribeirão do Pinhal|PR|4121901|17|
|37|Rio Azul|PR|4122008|19|
|38|Rio Branco do Sul|PR|4122206|18|
|39|Santo Antonio do Sudoeste|PR|4124400|58|
|40|São João|PR|4124806|48|
|41|São Jorge D Oeste|PR|4125209|163|
|42|São José das Palmeiras|PR|4125456|382|
|43|Sapopema|PR|4126207|113|
|44|Sengés|PR|4126306|61|
|45|Tijucas do Sul|PR|4127601|25|
|46|Toledo|PR|4127700|136|
|47|Turvo|PR|4127965|69|
|48|Ventania|PR|4128534|94|
|49|Verê|PR|4128609|72|
|50|Curiúva|PR|4107009|189|
|51|Foz do Jordão|PR|4108452|492|
|52|Ibiporã|PR|4109807|332|
|53|Marilândia do Sul|PR|4114906|750|
|54|Piên|PR|4119103|355|
|55|Porto Amazonas|PR|4120101|98|
|56|Prudentópolis|PR|4120606|28|


## Observações

- Caonsule sua prefeitura para detlhes de preenchimento do RPS


## Dependências

- PHP >= 7.1
- ext-curl
- ext-soap
- ext-zlib
- ext-dom
- ext-openssl
- ext-json
- ext-simplexml
- ext-libxml

### Outras Libs

- nfephp-org/sped-common
- justinrainbow/json-schema


## Contribuindo
Este é um projeto totalmente *OpenSource*, para usa-lo e modifica-lo você não paga absolutamente nada. Porém para continuarmos a mante-lo é necessário qua alguma contribuição seja feita, seja auxiliando na codificação, na documentação ou na realização de testes e identificação de falhas e BUGs.

**Este pacote está listado no [Packgist](https://packagist.org/)**

*Durante a fase de desenvolvimento e testes este pacote deve ser instalado com:*
```bash
composer require nfephp-org/sped-nfse-equiplano:dev-master
```

*Ou ainda,*
```bash
composer require nfephp-org/sped-nfse-equiplano:dev-master --prefer-dist
```

*Ou ainda alterando o composer.json do seu aplicativo inserindo:*
```json
"require": {
    "nfephp-org/sped-nfse-equiplano" : "dev-master"
}
```

> NOTA: Ao utilizar este pacote ainda na fase de desenvolvimento não se esqueça de alterar o composer.json da sua aplicação para aceitar pacotes em desenvolvimento, alterando a propriedade "minimum-stability" de "stable" para "dev".
> ```json
> "minimum-stability": "dev",
> "prefer-stable": true
> ```

*Após os stable realeases estarem disponíveis, este pacote poderá ser instalado com:*
```bash
composer require nfephp-org/sped-nfse-equiplano
```
Ou ainda alterando o composer.json do seu aplicativo inserindo:
```json
"require": {
    "nfephp-org/sped-sped-nfse-equiplano" : "^1.0"
}
```

## Forma de uso
vide a pasta *Examples*

## Log de mudanças e versões
Acompanhe o [CHANGELOG](CHANGELOG.md) para maiores informações sobre as alterações recentes.

## Testing

Todos os testes são desenvolvidos para operar com o PHPUNIT

## Security

Caso você encontre algum problema relativo a segurança, por favor envie um email diretamente aos mantenedores do pacote ao invés de abrir um ISSUE.

## Credits

Roberto L. Machado (owner and developer)

## License

Este pacote está diponibilizado sob LGPLv3 ou MIT License (MIT). Leia  [Arquivo de Licença](LICENSE.md) para maiores informações.

[ico-stable]: https://poser.pugx.org/nfephp-org/sped-nfse-equiplano/version
[ico-stars]: https://img.shields.io/github/stars/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-forks]: https://img.shields.io/github/forks/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-issues]: https://img.shields.io/github/issues/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/nfephp-org/sped-nfse-equiplano/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-version]: https://img.shields.io/packagist/v/nfephp-org/sped-nfse-equiplano.svg?style=flat-square
[ico-license]: https://poser.pugx.org/nfephp-org/nfephp/license.svg?style=flat-square
[ico-gitter]: https://img.shields.io/badge/GITTER-4%20users%20online-green.svg?style=flat-square


[link-packagist]: https://packagist.org/packages/nfephp-org/sped-nfse-equiplano
[link-travis]: https://travis-ci.org/nfephp-org/sped-nfse-equiplano
[link-scrutinizer]: https://scrutinizer-ci.com/g/nfephp-org/sped-nfse-equiplano/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/nfephp-org/sped-nfse-equiplano
[link-downloads]: https://packagist.org/packages/nfephp-org/sped-nfse-equiplano
[link-author]: https://github.com/nfephp-org
[link-issues]: https://github.com/nfephp-org/sped-nfse-equiplano/issues
[link-forks]: https://github.com/nfephp-org/sped-nfse-equiplano/network
[link-stars]: https://github.com/nfephp-org/sped-nfse-equiplano/stargazers


CHANGELOG.md
# Changelog

All Notable changes to `sped-nfse-equiplano` will be documented in this file.

Todas as atualizações a partir de 31/10/2018 devem observar os princípios [Mantendo o CHANGELOG](http://keepachangelog.com/).

## 1.0.0-dev 

## Notas da versão:

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


examples/jsonSchema_v1.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', 'On');
require_once '../bootstrap.php';

use JsonSchema\Constraints\Constraint;
use JsonSchema\Constraints\Factory;
use JsonSchema\SchemaStorage;
use JsonSchema\Validator;

$version = '1';

$jsonSchema = '{
    "title": "rps",
    "type": "object",
    "properties": {
        "nrrps": {
            "required": true,
            "type": "integer"
        },
        "nremissorrps": {
            "required": true,
            "type": "integer"
        },
        "dtemissaorps": {
            "required": true,
            "type": "string",
            "pattern": "^([0-9]{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])T(2[0-3]|[01][0-9]):[0-5][0-9]:[0-5][0-9])$"
        },
        "strps": {
            "required": true,
            "type": "integer",
            "minimum": 1,
            "maximum": 3
        },
        "tptributacao": {
            "required": true,
            "type": "integer",
            "minimum": 1,
            "maximum": 4
        },
        "isissretido": {
            "required": true,
            "type": "integer",
            "minimum": 1,
            "maximum": 2
        },
        "vltotalrps": {
            "required": true,
            "type": "number"
        },
        "vlliquidorps": {
            "required": true,
            "type": "number"
        },
        "tomador": {
            "required": false,
            "type": ["object","null"],
            "properties": {
                "nrdocumento": {
                    "required": false,
                    "type": ["string","null"],
                    "maxLength": 20
                },
                "tpdocumento": {
                    "required": false,
                    "type": ["integer","null"],
                    "minimum": 1,
                    "maximum": 3
                },
                "docestrangeiro": {
                    "required": false,
                    "type": ["string","null"],
                    "maxLength": 30
                },
                "nome": {
                    "required": true,
                    "type": "string",
                    "maxLength": 80
                },
                "email": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 6,
                    "maxLength": 60
                },
                "ie": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 1,
                    "maxLength": 20
                },
                "im": {
                    "required": false,
                    "type": ["string","null"],
                    "pattern": "^[0-9]{1,20}$"
                },
                "logradouro": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 0,
                    "maxLength": 40
                },
                "numero": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 0,
                    "maxLength": 10
                },
                "complemento": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 0,
                    "maxLength": 40
                },
                "bairro": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 0,
                    "maxLength": 30
                },
                "cidade": {
                    "required": false,
                    "type": ["string","null"],
                    "pattern": "^[0-9]{1,7}$"
                },
                "uf": {
                    "required": false,
                    "type": ["string","null"],
                    "maxLength": 2
                },
                "cidadeestrangeira": {
                    "required": false,
                    "type": ["string","null"],
                    "minLength": 0,
                    "maxLength": 30
                },
                "pais": {
                    "required": true,
                    "type": "string",
                    "maxLength": 40
                },
                "cep": {
                    "required": false,
                    "type": ["string","null"],
                    "pattern": "^[0-9]{8}$"
                },
                "telefone": {
                    "required": false,
                    "type": ["string","null"],
                    "maxLength": 20
                }
            }
        },
        "servico": {
            "required": true,
            "type": "object",
            "properties": {
                "nrservicoitem": {
                    "required": true,
                    "type": "integer"
                },    
                "nrservicosubitem": {
                    "required": true,
                    "type": "integer"
                },
                "vlservico": {
                    "required": true,
                    "type": "number"
                },
                "vlaliquota": {
                    "required": true,
                    "type": "number"
                },
                "vlbasecalculo": {
                    "required": true,
                    "type": "number"
                },
                "vlissservico": {
                    "required": true,
                    "type": "number"
                },
                "discriminacao": {
                    "required": true,
                    "type": "string",
                    "maxLength": 1024
                },
                "deducao": {
                    "required": false,
                    "type": ["object","null"],
                    "properties": {
                        "vldeducao": {
                            "required": true,
                            "type": "number"
                        },
                        "justificativa": {
                            "required": true,
                            "type": "string",
                            "maxLength": 255
                        }
                    }
                }
            }
        },    
        "retencoes": {
            "required": false,
            "type": ["object","null"],
            "properties": {
                "vlcofins": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlcsll": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlinss": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlirrf": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlpis": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vliss": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlaliquotacofins": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlaliquotacsll": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlaliquotainss": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlaliquotairrf": {
                    "required": false,
                    "type": ["number","null"]
                },
                "vlaliquotapis": {
                    "required": false,
                    "type": ["number","null"]
                } 
            }    
        }
    }
}';


$std = new \stdClass();
$std->nrversaoxml = 1;
$std->nrrps = 1;
$std->nremissorrps = 1;
$std->dtemissaorps = '2019-08-08T14:58:14';
$std->strps = 1; //1=converter, 2=converter e cancelar NFS, 3=cancelar RPS
$std->tptributacao = 1; //1=tributado no munícipio, 2=em outro munícipio, 3=isento/imune, 4=suspenso/decisão judicial
$std->isissretido = 2; //1=sim, 2=não
$std->vltotalrps = 100.00;
$std->vlliquidorps = 100.00;

$std->tomador = new \stdClass();
$std->tomador->nrdocumento = '12345678901234';
$std->tomador->tpdocumento = 2; //1=cpf, 2=cnpj, 3=estrangeiro
$std->tomador->documentoestrangeiro = null; //obrigatório se tpDocumento=3 max 30 caracteres
$std->tomador->nome = 'Fulano da Tal';
$std->tomador->email = null;
$std->tomador->ie = null;
$std->tomador->im = null;
$std->tomador->logradouro = null;
$std->tomador->numero = null;
$std->tomador->complemento = null;
$std->tomador->bairro = null;
$std->tomador->cidade = null;
$std->tomador->uf = null;
$std->tomador->cidadeestrangeira = null;
$std->tomador->pais = 'BRASIL';
$std->tomador->cep = null;
$std->tomador->telefone = null;

$std->servico = new \stdClass();
$std->servico->nrservicoitem = 4;
$std->servico->nrservicosubitem = 7;
$std->servico->vlservico = 100.00;
$std->servico->vlaliquota = 0.5;
$std->servico->vlbasecalculo = 100.00;
$std->servico->vlissservico = 5.00;
$std->servico->discriminacao = 'Teste de Emissao';

$std->servico->deducao = new \stdClass();
$std->servico->deducao->vldeducao = 12;
$std->servico->deducao->justificativa = 'Sem justificativa';

$std->retencoes = new \stdClass();
$std->retencoes->vlcofins = 0;
$std->retencoes->vlcsll = 0;
$std->retencoes->vlinss = 0;
$std->retencoes->vlirrf = 0;
$std->retencoes->vlpis = 0;
$std->retencoes->vliss = 0;
$std->retencoes->vlaliquotacofins = 0;
$std->retencoes->vlaliquotacsll = 0;
$std->retencoes->vlaliquotainss = 0;
$std->retencoes->vlaliquotairrf = 0;
$std->retencoes->vlaliquotapis = 0;

// Schema must be decoded before it can be used for validation
$jsonSchemaObject = json_decode($jsonSchema);
if (empty($jsonSchemaObject)) {
    echo "<h2>Erro de digitação no schema ! Revise</h2>";
    echo "<pre>";
    print_r($jsonSchema);
    echo "</pre>";
    die();
}

// The SchemaStorage can resolve references, loading additional schemas from file as needed, etc.
$schemaStorage = new SchemaStorage();
// This does two things:
// 1) Mutates $jsonSchemaObject to normalize the references (to file://mySchema#/definitions/integerData, etc)
// 2) Tells $schemaStorage that references to file://mySchema... should be resolved by looking in $jsonSchemaObject
$schemaStorage->addSchema('file://mySchema', $jsonSchemaObject);
// Provide $schemaStorage to the Validator so that references can be resolved during validation
$jsonValidator = new Validator(new Factory($schemaStorage));
// Do validation (use isValid() and getErrors() to check the result)
$jsonValidator->validate(
    $std,
    $jsonSchemaObject,
    Constraint::CHECK_MODE_COERCE_TYPES  //tenta converter o dado no tipo indicado no schema
);

if ($jsonValidator->isValid()) {
    echo "The supplied JSON validates against the schema.<br/>";
} else {
    echo "Dados não validados. Violações:<br/>";
    foreach ($jsonValidator->getErrors() as $error) {
        echo sprintf("[%s] %s<br/>", $error['property'], $error['message']);
    }
    die;
}

//salva se sucesso
file_put_contents("../storage/jsonSchemes/v$version/rps.schema", $jsonSchema);